Java: Add support for algorithm names specified in `.properties` files to `java/potentially-weak-cryptographic-algorithm`

egregius313 opened the pull request:
This adds a new query to detect uses of weak hashing algorithms, where the algorithm name comes from a `.properties` file.

> [!IMPORTANT]
> On hold until https://github.com/github/codeql/pull/14802 is merged.

```[tasklist]
- [ ] https://github.com/github/codeql/pull/14802
- [x] Convert integration test to query test
```

atorralba commented on java/ql/integration-tests/all-platforms/java/properties-files/weak-hashing/src/main/java/com/example/WeakHashing.java:
```diff
@@ -0,0 +1,16 @@
+package com.example;
+
+import java.utils.Properties;
```
Your integration test is failing because of this import:

```
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild] > Task :compileJava FAILED
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild] (...) ql/java/ql/integration-tests/all-platforms/java/properties-files/weak-hashing/src/main/java/com/example/WeakHashing.java:3: error: package java.utils does not exist
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild] import java.utils.Properties;
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild]                  ^
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild] 1 error
[2023-08-23 18:21:09] [build-stdout] [2023-08-23 18:21:09] [autobuild] FAILURE: Build failed with an exception.
```

It should be `java.util.Properties`.

```suggestion
import java.util.Properties;
```

jcogs33 reviewed (commented):
I have a few high-level questions for my own understanding of this query (not a full review):
 - Does this query only cover weak hashing or also weak encryption? (I'm confused by the query name which only mentions "hashing" versus the description which seems to imply "cryptographic algorithms" in general).
 - If the query is intended to only cover hashing, should it have its own QHelp file instead of copying the `BrokenCryptoAlgorithm.qhelp`? I found it confusing that a QHelp file titled "Weak Hashing " was mentioning encryption algorithms.
 -  Out of curiosity, what is the reason why this needs to be a separate query instead of being included as another source in the `java/weak-cryptographic-algorithm` query  (and/or in `java/potentially-weak-cryptographic-algorithm`)?

egregius313 commented:
@jcogs33 

> Does this query only cover weak hashing or also weak encryption? (I'm confused by the query name which only mentions "hashing" versus the description which seems to imply "cryptographic algorithms" in general).

I've updated the wording in the query and the PR. This is only officially meant to cover weak hashing. At the moment it may have results for both, since it relies on `CryptoAlgoSpec`. If we eventually split `CryptoAlgoSpec` into encryption and hashing separately, then this would only use the hashing version. (Splitting them now would likely cause too much disruption in alerts).

> If the query is intended to only cover hashing, should it have its own QHelp file instead of copying the BrokenCryptoAlgorithm.qhelp? I found it confusing that a QHelp file titled "Weak Hashing " was mentioning encryption algorithms.

Yes, that was an oversight from when I first made the query. It makes more sense to have its own, which I have added.

> Out of curiosity, what is the reason why this needs to be a separate query instead of being included as another source in the java/weak-cryptographic-algorithm query (and/or in java/potentially-weak-cryptographic-algorithm)?

This is because since this is meant to cover weak hashing algorithms, it makes more sense to be placed under CWE-328, instead of 327.

jcogs33 commented:
> This is only officially meant to cover weak hashing.

What is the reason why this query needs to be specific to hashing and cannot include encryption as well? (I think I'm missing some context. Are encryption algorithms never specified in `.properties` files?)

> At the moment it may have results for both, since it relies on CryptoAlgoSpec.

Is there any concern that this query will have too many FPs from weak encryption as a result of using `CryptoAlgoSpec`? Did FPs of this sort come up when testing? If I ran a query titled "Weak Hashing ", but received results related to encryption, I would be very confused. :smile:

atorralba commented on java/ql/lib/semmle/code/java/security/MaybeBrokenCryptoAlgorithmQuery.qll:
```diff
@@ -34,11 +37,44 @@ private predicate objectToString(MethodCall ma) {
+  }
+}
+
+/**
+ * Get the string value represented by the given expression.
+ *
+ * If the value is a string literal, return the literal value.
+ * If the value is a call to `java.util.Properties::getProperty`, return the potential values of the property.
+ */
+string insecureAlgorithmName(DataFlow::Node algo) {
```
Sorry, just a quick thing before approving:

This predicate should be named differently, since it does nothing related to `insecure`, it just gets the string value of the node `algo` if it's either a `StringLiteral` or a `GetPropertyMethodCall`. So I'd suggest something like:

```suggestion
/**
 * Gets the string value represented by the given expression node.
 *
 * If the value is a string literal, gets the literal value.
 * If the value is a call to `java.util.Properties::getProperty`, gets the potential values of the property.
 */
string getStringValue(DataFlow::ExprNode algo) {
```

atorralba replied:
(Principle being name things for what they do, not for what they'll be used for)

atorralba commented on java/ql/src/Security/CWE/CWE-327/MaybeBrokenCryptoAlgorithm.ql:
```diff
@@ -13,16 +13,27 @@
+/**
+ * Get the string value represented by the given expression.
+ *
+ * If the value is a string literal, return the literal value.
+ * If the value is a call to `java.util.Properties::getProperty`, return the potential values of the property.
```
```suggestion
 * If the value is a string literal, get the literal value.
 * If the value is a call to `java.util.Properties::getProperty`, get the potential values of the property.
```

atorralba reviewed (approved):
🎉
